ui/core/src/utils/transform-data.ts

```diff
@@ -0,0 +1,212 @@
+// Copyright 2024 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  JoinByColumnValueTransformSpec,
+  MergeColumnsTransformSpec,
+  MergeIndexedColumnsTransformSpec,
+  Transform,
+} from '@perses-dev/core';
+import { useMemo } from 'react';
+
+// export function applyJoinTransform(
+//   data: Array<Record<string, unknown>>,
+//   transform: Transform<JoinByColumnValueTransformSpec>
+// ): Array<Record<string, unknown>> {
+//   // If column is undefined or empty, return data as is
+//   if (!transform.spec.plugin.spec.column) {
+//     return data;
+//   }
+//
+//   const column: string = transform.spec.plugin.spec.column;
+//   const entriesHashed: { [key: string]: Record<string, unknown> } = {};
+//
+//   for (const entry of data) {
+//     if (Object.keys(entry).includes(column)) {
+//       const hash = String(entry[column]);
+//       const entryHashed = entriesHashed[hash];
+//
+//       if (entryHashed) {
+//         entriesHashed[hash] = { ...entryHashed, ...entry };
+//       } else {
+//         entriesHashed[hash] = { ...entry };
+//       }
+//     }
+//   }
+//   return Object.values(entriesHashed);
+// }
+
+export function applyJoinTransform(
```

Comment:
The nice visual examples you put in the PR description, maybe you could add them as comment to these functions? You could use http://markdowntable.com/ if needed 🤓

Reply:
Yup, good idea